compiler/rustc_typeck/src/check/gather_locals.rs

```diff
@@ -7,6 +7,31 @@ use rustc_middle::ty::Ty;
 use rustc_span::Span;
 use rustc_trait_selection::traits;
 
+/// A declaration is an abstraction of [hir::Local] and [hir::Let].
+///
+/// It must have a hir_id, as this is how we connect gather_locals to the check functions.
+pub(super) struct Declaration<'a> {
+    pub hir_id: hir::HirId,
+    pub pat: &'a hir::Pat<'a>,
+    pub ty: Option<&'a hir::Ty<'a>>,
+    pub span: Span,
+    pub init: Option<&'a hir::Expr<'a>>,
+}
```

Comment:
Addition of this further suggests to me that the HIR representation of `let-else` should have been modeled after a `hir::Local` and not `if-let`… Out of scope for this PR though.

Reply:
At least we could pretty easily add support for type annotations to `if-let`, i.e. this which you can't do today:

```rust
if let Some(x): &mut i32 = optional() { ... }
```

If anything, of all the name bindings in the language, `hir::Local` is the odd one out, having an optional init expression.

Reply:
I guess a more perfect refactoring is to extract the common parts (excluding the ty/option ty) into a `hir::Decl` or something. But similarly this is a big enough change already without doing to `hir::Local` what I've done to let-else.

Ultimately it's just not that big a deal for the codebase. The other thing that's apparent from `gather_locals.rs` is that we only needed common functionality in one place and the rest of a change like that would just be busywork.